### PR TITLE
[Core] refactor `transformer_2d` forward logic into meaningful conditions.

### PR DESCRIPTION
**Pull request by sayakpaul**
# What does this PR do?

Refactors the `forward()` of `Transformers2DModel` for easier readability. 

More specifically, the PR refactors the `forward()` method of `Transformers2DModel` to have the following unified structure:

```python
def forward(self):
    # handle attention masking. 
    ...

    # 1. input-level operations.
    if self.is_input_continuous:
        ... = self._operate_on_continuous_inputs(...)
    elif self.is_input_vectorized:
    	... = self.latent_image_embedding(...) # for vectorized inputs only this is needed.
    elif self.is_input_patches:
        ... = self._operate_on_patched_inputs(...)

    # 2. Blocks (common across all the variants)
    for block in self.transformer_blocks:
    	...

    # 3. Outputs.
    if self.is_input_continuous:
        ... = self._get_output_for_continuous_inputs(...)
    elif self.is_input_vectorized:
        output = self._get_output_for_vectorized_inputs(...)
    elif self.is_input_patches:
        output = self._get_output_for_patched_inputs(...)

    return output
```

About the possibility of spinning out separate transformer classes for patched and vectorized inputs, what's the consensus? Currently, we have DiT and PixArt-Alpha that use patched inputs. So, for those checkpoints ([DiT](https://huggingface.co/facebook/DiT-XL-2-256) and [PixArt-Alpha](https://huggingface.co/PixArt-alpha/PixArt-XL-2-1024-MS)), are we thinking of just updating the configs to use `PatchedTransformer2DModel`? (Same could be done for the `VectorizedTransformer2DModel`)

If that's the case I think it could be quite breaking of a change. Many folks use the DiT and PixArt models especially in the light of many Open SoRA initiative. Introducing a change like this could be problematic. Should we consider throwing a deprecation warning, instead? 

@yiyixuxu @DN6 please let me know your thoughts.

### CODE REVIEW

**Comment by sayakpaul** [comment]
@DN6 @yiyixuxu a gentle ping.

File: src/diffusers/models/transformers/transformer_2d.py [hunk 1]
```diff
@@ -402,41 +402,18 @@ def forward(
 
         # 1. Input
         if self.is_input_continuous:
-            batch, _, height, width = hidden_states.shape
+            batch_size, _, height, width = hidden_states.shape
```
Comment (DN6) [inline src/diffusers/models/transformers/transformer_2d.py:405]: Think this shape check can be moved out of the if statements safely .e.g 

```suggestion
        batch_size, _, height, width = hidden_states.shape

        if self.is_input_continuous:
            residual = hidden_states
            hidden_states, inner_dim = self._operate_on_continuous_inputs(hidden_states)
        elif self.is_input_vectorized:
            hidden_states = self.latent_image_embedding(hidden_states)
        elif self.is_input_patches:
            height, width = height // self.patch_size, width // self.patch_size
            hidden_states, encoder_hidden_states, timestep, embedded_timestep = self._operate_on_patched_inputs(
                hidden_states, encoder_hidden_states, timestep, added_cond_kwargs
            )

```

**Review by DN6 — APPROVED** [review]
LGTM. Left a comment, but it's mostly a nit.

**Comment by sayakpaul** [comment]
@DN6 WDYT?

> About the possibility of spinning out separate transformer classes for patched and vectorized inputs, what's the consensus? Currently, we have DiT and PixArt-Alpha that use patched inputs. So, for those checkpoints ([DiT](https://huggingface.co/facebook/DiT-XL-2-256) and [PixArt-Alpha](https://huggingface.co/PixArt-alpha/PixArt-XL-2-1024-MS)), are we thinking of just updating the configs to use PatchedTransformer2DModel? (Same could be done for the VectorizedTransformer2DModel)

(refer to the OP)

**Comment by yiyixuxu** [comment]
I'm cool with this PR once you and @DN6 are happy with this!


for this, yes, I agree it is a highly used block, and I think we should give more thought to avoid breaking changes. can potentially deprecate `Tranformer2DModel` and use 3 different `Transformers2DModel` instead

> If that's the case I think it could be quite breaking of a change. Many folks use the DiT and PixArt models especially in the light of many Open SoRA initiative. Introducing a change like this could be problematic. Should we consider throwing a deprecation warning, instead?

**Comment by sayakpaul** [comment]
Cool. Will work on the deprecation in a future PR then as we discuss the best design.

**Comment by sayakpaul** [comment]
@DN6 I had to revert your suggestion because `height` and `width` are calculated differently when inputs are continuous and patched. Hope that's okay. 

Gonna merge the PR after the CI is green.

**Pull request merged**